executor/cluster_reader.go

```diff
@@ -215,3 +218,231 @@ func (e *clusterConfigRetriever) retrieve(ctx sessionctx.Context) ([][]types.Dat
+		typ := srv.ServerType
+		// Skip some node type which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_log WHERE type='tikv'
+		if len(e.extractor.NodeTypes) > 0 && !e.extractor.NodeTypes.Exist(typ) {
+			continue
+		}
+		address := srv.Address
+		// Skip some node address which has been filtered in WHERE cluase
+		// e.g: SELECT * FROM cluster_log WHERE address='192.16.8.12:2379'
+		if len(e.extractor.Addresses) > 0 && !e.extractor.Addresses.Exist(address) {
+			continue
+		}
+		statusAddr := srv.StatusAddr
+		if len(statusAddr) == 0 {
+			ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("%s node %s does not contain status address", typ, address))
+			continue
+		}
+		ch := make(chan logStreamResult)
+		results = append(results, ch)
+
+		go func(ch chan logStreamResult, serverType, address, statusAddr string) {
+			util.WithRecovery(func() {
+				defer close(ch)
+
+				// The TiDB provide diagnostics service via status address
+				remote := address
+				if serverType == "tidb" {
+					remote = statusAddr
+				}
+				conn, err := grpc.Dial(remote, opt)
+				if err != nil {
+					ch <- logStreamResult{addr: address, typ: serverType, err: err}
+					return
+				}
+				defer terror.Call(conn.Close)
+
+				cli := diagnosticspb.NewDiagnosticsClient(conn)
+				stream, err := cli.SearchLog(context.Background(), req)
+				if err != nil {
+					ch <- logStreamResult{addr: address, typ: serverType, err: err}
+					return
+				}
+
+				for {
+					res, err := stream.Recv()
+					if err != nil && err == io.EOF {
+						return
+					}
+					if err != nil {
+						ch <- logStreamResult{addr: address, typ: serverType, err: err}
+						return
+					}
+					ch <- logStreamResult{next: ch, addr: address, typ: serverType, messages: res.Messages}
+				}
+			}, nil)
+		}(ch, typ, address, statusAddr)
+	}
+
+	e.results = results
+	return nil
+}
+
+func (e *clusterLogRetriever) retrieve(ctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.isDrained {
+		return nil, nil
+	}
+
+	if !e.retrieving {
+		e.retrieving = true
+		if err := e.startRetrieving(ctx); err != nil {
+			e.isDrained = true
+			return nil, err
+		}
+
+		// initialize the heap
+		e.heap = &logResponseHeap{}
+		n := len(e.results)
+		for i := n - 1; i >= 0; i-- {
+			result := <-e.results[i]
+			if result.err != nil || len(result.messages) == 0 {
+				if result.err != nil {
+					ctx.GetSessionVars().StmtCtx.AppendWarning(result.err)
+				}
+				e.results = append(e.results[:i], e.results[i+1:]...)
+				continue
+			}
+			*e.heap = append(*e.heap, result)
+		}
+		heap.Init(e.heap)
+	}
+
+	// Merge the results
+	var finalRows [][]types.Datum
+	for e.heap.Len() > 0 && len(finalRows) < 1024 {
+		min := heap.Pop(e.heap).(logStreamResult)
+		item := min.messages[0]
+		time := time.Unix(item.Time/1000, (item.Time%1000)*int64(time.Millisecond))
+		finalRows = append(finalRows, types.MakeDatums(
+			time.Format("2006/01/02 15:04:05.000"),
```

Comment:
I don't understand...

Reply:
Make the result time has the same format with original log text but don't contain the time zone information.

Reply:
I suggest rename `time`. It overwrites package name `time`.

Reply:
done